packages/fxa-auth-server/lib/routes/subscriptions/paypal.ts

```diff
@@ -89,66 +90,93 @@ export class PayPalHandler extends StripeWebhookHandler {
-      // on record.  But it's a bug with the frontend if it happens and we
-      // shouldn't allow it.
-      throw error.billingAgreementExists(customer.id);
-    }
+      const isPaypalCustomer = hasPaypalSubscription(customer);
+      const { token, metricsContext } = request.payload as Record<
+        string,
+        string
+      >;
+      const currentBillingAgreement =
+        this.stripeHelper.getCustomerPaypalAgreement(customer);
+
+      // In theory the frontend should handle this state upon the customer
+      // response.  The customer should fix their payment method before we can
+      // allow any additional subscriptions.
+      if (isPaypalCustomer && !currentBillingAgreement) {
+        throw error.missingPaypalBillingAgreement(customer.id);
+      }
+      if (!isPaypalCustomer && !token) {
+        throw error.missingPaypalPaymentToken(customer.id);
+      }
+      if (isPaypalCustomer && token) {
+        // This seems unfriendly to users since we could just proceed with the BA
+        // on record.  But it's a bug with the frontend if it happens and we
+        // shouldn't allow it.
+        throw error.billingAgreementExists(customer.id);
+      }
 
-    const { sourceCountry, subscription } = !!token
-      ? await this._createPaypalBillingAgreementAndSubscription({
-          request,
-          uid,
-          customer,
-        })
-      : await this._createPaypalSubscription({ request, customer });
+      const { sourceCountry, subscription } = !!token
+        ? await this._createPaypalBillingAgreementAndSubscription({
+            request,
+            uid,
+            customer,
+          })
+        : await this._createPaypalSubscription({ request, customer });
 
-    await this.customerChanged(request, uid, email);
+      await this.customerChanged(request, uid, email);
 
-    this.log.info('subscriptions.createSubscriptionWithPaypal.success', {
-      uid,
-      subscriptionId: subscription.id,
-    });
+      this.log.info('subscriptions.createSubscriptionWithPaypal.success', {
+        uid,
+        subscriptionId: subscription.id,
+      });
 
-    await sendFinishSetupEmailForStubAccount({
-      uid,
-      account,
-      subscription,
-      stripeHelper: this.stripeHelper,
-      mailer: this.mailer,
-      subscriptionAccountReminders: this.subscriptionAccountReminders,
-      metricsContext,
-    });
+      await sendFinishSetupEmailForStubAccount({
+        uid,
+        account,
+        subscription,
+        stripeHelper: this.stripeHelper,
+        mailer: this.mailer,
+        subscriptionAccountReminders: this.subscriptionAccountReminders,
+        metricsContext,
+      });
 
-    return {
-      sourceCountry,
-      subscription: filterSubscription(subscription),
-    };
+      return {
+        sourceCountry,
+        subscription: filterSubscription(subscription),
+      };
+    } catch (err) {
+      try {
```

Comment:
Since we are duplicating this logic for paypal and stripe, could we move it into a common function or method, maybe in `fxa-shared`?

Reply:
I've spent roughly 2 hours trying to refactor this this morning, the problem ends up going down a few different paths: we have a few different testing strategies that we employ, resulting from a few different design patterns, with how I structured things originally to move that `deleteAccountIfUnverified` function to its own module, if I want to add a wrapper in there, it becomes really difficult to stub the first function when you want to test the logic of the wrapped function (e.g the check for `verifierSetAt` and ignoring certain errors). We could either restructure the tests (which is time intensive) or look to employ some other way of structurng that module (exporting a factory that has those methods) but then it is going require changes to the return signatures of the methods that invoke them which will require a bunch of refactoring of those tests.

My vote is to leave this as is even though its not as DRY as I'd like, it has tests in place that will help ensure that if that logic is changed from whats intended we will catch any errors and it spares us from having to either do more refactoring and increasing the surface area of what this touches or employing more patterns that I don't really see us use elsewhere (which would likely be more of a burden to maintain than the fact these checks are implemented twice).